### PR TITLE
Add ErrorSuppressionFixer

### PR DESCRIPTION
**Pull request by kubawerlos**
#3035

Initial commit - fixer and tests are based on `SilencedDeprecationErrorFixer` thus @HeahDude is added as author.

Option allows only boolean value meaning to do what is should or not.

Having also `null` to ignore option would make `false` doing the opposite to `true`: 
- for `silence_deprecation_error` it would sense - to remove `@` from deprecation notices
- but for `noise_remaining_usages` what would that mean? Add `@` everywhere?

Feel free to propose better naming for options, descriptions and interesting test cases.

My doubt is line 106 in `ErrorSuppressionFixer.php` - I thought it should be something like:

```php
$indexes[$prev] = (true === $this->configuration['noise_remaining_usages']) ? false : null;
```

but that failed some tests. Anyway it's just a hint where to look for flaws.

### CODE REVIEW

File: src/Fixer/LanguageConstruct/ErrorSuppressionFixer.php [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+use PhpCsFixer\FixerConfiguration\FixerConfigurationResolver;
+use PhpCsFixer\FixerConfiguration\FixerOptionBuilder;
+use PhpCsFixer\FixerDefinition\CodeSample;
+use PhpCsFixer\FixerDefinition\FixerDefinition;
+use PhpCsFixer\Tokenizer\Token;
+use PhpCsFixer\Tokenizer\Tokens;
+
+/**
+ * @author Jules Pietri <jules@heahprod.com>
+ * @author Kuba Werłos <werlos@gmail.com>
```
Comment (keradus) [inline src/Fixer/LanguageConstruct/ErrorSuppressionFixer.php]: damn, I anticipated `ś` at the end ;)

**Comment by keradus** [comment]
@TomasVotruba , could you please help our community once more ?
This PR needs a decent review for quite a long already. Not only the code needs it, but also the idea itself.
Could you please take care of it ?

**Comment by kubawerlos** [comment]
@keradus it needs update - I see test are not covering all cases

File: src/Fixer/LanguageConstruct/ErrorSuppressionFixer.php [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+            'Adding or removing ``@`` might cause changes to code behaviour.'
+        );
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function isCandidate(Tokens $tokens)
+    {
+        return $tokens->isTokenKindFound(T_STRING);
```
Comment (TomasVotruba) [inline src/Fixer/LanguageConstruct/ErrorSuppressionFixer.php]: I'd check for these too:

- ','
- `T_CONSTANT` (not sure if that covers E_USER_DEPRECATED),

File: src/Fixer/LanguageConstruct/ErrorSuppressionFixer.php [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+     * {@inheritdoc}
+     */
+    protected function applyFix(\SplFileInfo $file, Tokens $tokens)
+    {
+        $indexes = [];
+
+        for ($index = $tokens->count() - 1; $index >= 0; --$index) {
+            $token = $tokens[$index];
+
+            if ($token->equals('@') && true === $this->configuration['noise_remaining_usages']) {
```
Comment (TomasVotruba) [inline src/Fixer/LanguageConstruct/ErrorSuppressionFixer.php]: First, there should be check for token, as we don't care about any other

```php
if (! $token->isGivenKind(T_STRING)) {
    continue;
}
```

File: src/Fixer/LanguageConstruct/ErrorSuppressionFixer.php [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+            if ($tokens[$prev]->isGivenKind(T_STRING) || $tokens[$prev]->equals('@')) {
+                $indexes[$prev] = null;
+
+                continue;
+            }
+
+            $end = $tokens->findBlockEnd(Tokens::BLOCK_TYPE_PARENTHESIS_BRACE, $tokens->getNextTokenOfKind($index, [T_STRING, '(']));
+            if ($tokens[$tokens->getPrevMeaningfulToken($end)]->equals([T_STRING, 'E_USER_DEPRECATED'])) {
+                $indexes[$start] = (true === $this->configuration['silence_deprecation_error']) ? true : null;
+            }
```
Comment (TomasVotruba) [inline src/Fixer/LanguageConstruct/ErrorSuppressionFixer.php]: Codeblock from `if (!$token->equals([T_STRING, 'trigger_error'], false)) {` to here should be decoupled to own method to improve readability and maintainability.

E.g. `$this->fixTriggerError($index, $tokens)`

File: src/Fixer/LanguageConstruct/ErrorSuppressionFixer.php [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+
+            if ($tokens[$prev]->isGivenKind(T_STRING) || $tokens[$prev]->equals('@')) {
+                $indexes[$prev] = null;
+
+                continue;
+            }
+
+            $end = $tokens->findBlockEnd(Tokens::BLOCK_TYPE_PARENTHESIS_BRACE, $tokens->getNextTokenOfKind($index, [T_STRING, '(']));
+            if ($tokens[$tokens->getPrevMeaningfulToken($end)]->equals([T_STRING, 'E_USER_DEPRECATED'])) {
+                $indexes[$start] = (true === $this->configuration['silence_deprecation_error']) ? true : null;
```
Comment (TomasVotruba) [inline src/Fixer/LanguageConstruct/ErrorSuppressionFixer.php]: Ternary operator with composed conditions is hard to read, yoda with moreover.
What does this do?

Could you split it to or simplify? Well, comparison bool to `true` could shortened:

```php
$indexes[$start] = $this->configuration['silence_deprecation_error'] ? true : null;
```

File: src/Fixer/LanguageConstruct/ErrorSuppressionFixer.php [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+            }
+        }
+
+        foreach ($indexes as $key => $index) {
+            if (true === $index) {
+                $tokens->insertAt($key, new Token('@'));
+            } elseif (false === $index) {
+                $tokens->clearAt($key);
+            }
+        }
```
Comment (TomasVotruba) [inline src/Fixer/LanguageConstruct/ErrorSuppressionFixer.php]: What does this do? Could you maybe decouple to own method with descriptive name?

**Comment by TomasVotruba** [comment]
@keradus Sure I'll give it some time, when you asked so nicely :+1: 

Personally I'd split it to fixers, since:

- one feature is `trigger_error` `E_USER_DEPRECATED` with `@`
- second is remove (almost) all `@` 
    - here should be option to whitelist functions to skip with `trigger_error` as default; also many people use `@mkdir` for checking already existing directory etc.

File: src/Fixer/LanguageConstruct/ErrorSuppressionFixer.php [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+
+            if ($token->equals('@') && true === $this->configuration['noise_remaining_usages']) {
+                if (!array_key_exists($index, $indexes)) {
+                    $indexes[$index] = false;
+
+                    continue;
+                }
+            }
+
+            if (!$token->equals([T_STRING, 'trigger_error'], false)) {
```
Comment (TomasVotruba) [inline src/Fixer/LanguageConstruct/ErrorSuppressionFixer.php]: Here would be best place to check against list of allowed functions

**Comment by TomasVotruba** [comment]
I've just read original issue and discussion.
If I understand that correctly, this should cover only the 2nd fixer is suggested.

So this fixer should only take care of  `@` removal with `trigger_error` function as default.

**Comment by keradus** [comment]
that's the thing, we can not have 2 fixers, one adding `@` (even only in some cases), one removing `@` , as they will be conflicting, and whole PHP CS Fixer project is written in a way to avoid conflicting rules.

**Comment by TomasVotruba** [comment]
The other case is to keep this fixer as it is and accept it will containt this hidden reference to other fixer - "we skip trigger_error here, because it is covered by @see AnotherFixer".

That would definitelly made my code review less wtf.

File: src/Fixer/LanguageConstruct/ErrorSuppressionFixer.php (hunk 2 above)
Comment (kubawerlos) [reply src/Fixer/LanguageConstruct/ErrorSuppressionFixer.php]: For files with no constants we still might want to remove `@`.

File: src/Fixer/LanguageConstruct/ErrorSuppressionFixer.php (hunk 3 above)
Comment (kubawerlos) [reply src/Fixer/LanguageConstruct/ErrorSuppressionFixer.php]: `@` has no kind :( only we can do is `->equals('@')`.

**Comment by kubawerlos** [comment]
@TomasVotruba thanks for the review, I've reworked fixer, it should be much easier to read it's code. Also older version was missing some cases.

File: README.rst [hunk 8]
```diff
@@ -553,6 +553,20 @@ Choose from the list of available rules:
...
+  Error control operator should be added to deprecation notices and/or
+  removed from other cases.
+
+  *Risky rule: adding and/or removing ``@`` might cause changes to code behaviour.*
+
+  Configuration options:
+
+  - ``mute_deprecation_error`` (``false``, ``true``): whether to add ``@`` in deprecation
+    notices; defaults to ``true``
+  - ``noise_remaining_usages`` (``false``, ``true``): whether to remove ``@`` in remaining
```
Comment (keradus) [inline README.rst]: If I would only read option, then I understand `noise_remaining_usages=yes` to actually add `@`.
Is it really adding them ?

Comment (kubawerlos) [reply README.rst]: How come? `@` is known a silence operator, so "noising" would naturally mean removing it?

I've chosen the names "mute" and "noise" as they are opposite words and to have an order (which is alphabetical by option name) where the "remaining" option is second.

Comment (keradus) [reply README.rst]: my bad, ignore :+1:

**Comment by julienfalque** [comment]
Is there a specific reason not to implement some `exclude` option as proposed in #3035? IMO it's a must have, as some `@` usages are legit, e.g.:
```php
// failure is handled so one doesn't care about the E_WARNING
if (false === $file = @fopen(...)) {
    // throw some exception
}
```

**Comment by keradus** [comment]
ping @kubawerlos 
what is the state of this PR ?

File: src/Fixer/LanguageConstruct/ErrorSuppressionFixer.php [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function createConfigurationDefinition()
+    {
+        return new FixerConfigurationResolver([
+            (new FixerOptionBuilder(self::OPTION_MUTE_DEPRECATION_ERROR, 'Whether to add `@` in deprecation notices.'))
+                ->setAllowedValues([true, false])
```
Comment (ntzm) [inline src/Fixer/LanguageConstruct/ErrorSuppressionFixer.php]: Should be `->setAllowedTypes(['bool'])` instead

File: src/Fixer/LanguageConstruct/ErrorSuppressionFixer.php [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+        ]);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function applyFix(\SplFileInfo $file, Tokens $tokens)
+    {
+        $functionsAnalyzer = new FunctionsAnalyzer();
+        $excludedFunctions = array_map('strtolower', $this->configuration[self::OPTION_NOISE_REMAINING_USAGES_EXCLUDE]);
```
Comment (ntzm) [inline src/Fixer/LanguageConstruct/ErrorSuppressionFixer.php]: Please use

```php
array_map(function ($function) {
    return strtolower($function);
}, $this->configuration[self::OPTION_NOISE_REMAINING_USAGES_EXCLUDE]);
```

instead, see #3769

File: src/Fixer/LanguageConstruct/ErrorSuppressionFixer.php [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+                new CodeSample(
+                    "<?php\n@mkdir(\$dir);\n@unlink(\$path);\n",
+                    [
+                        self::OPTION_NOISE_REMAINING_USAGES => true,
+                        self::OPTION_NOISE_REMAINING_USAGES_EXCLUDE => ['unlink'],
+                    ]
+                ),
+            ],
+            null,
+            'Adding and/or removing `@` might cause changes to code behaviour.'
```
Comment (ntzm) [inline src/Fixer/LanguageConstruct/ErrorSuppressionFixer.php]: Is also risky if `trigger_error` is overwritten

**Comment by keradus** [comment]
Thank you @kubawerlos.

**Pull request merged**